elrond-wasm/src/types/interaction/contract_call.rs

```diff
@@ -102,21 +122,27 @@ where
     /// If this is an ESDT call, it converts it to a regular call to ESDTTransfer.
     /// Async calls require this step, but not `transfer_esdt_execute`.
     fn convert_to_esdt_transfer_call(self) -> Self {
-        if self.payment_token.is_egld() {
+        if self.payments.len() == 1 {
+            self.convert_to_single_transfer_esdt_call()
+        } else {
+            self.convert_to_multi_transfer_esdt_call()
+        }
+    }
+
+    fn convert_to_single_transfer_esdt_call(self) -> Self {
```

Comment:
Just for readability purposes, I think it would be better to have something like `let payment = &self.payments[0];` and use that instead of indexing everywhere. Way less prone to mistakes too.